src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampToken.cs

```diff
@@ -0,0 +1,531 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Pkcs.Asn1;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography.Xml;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    public sealed class Rfc3161TimestampToken
+    {
+        private SignedCms _parsedDocument;
+
+        public Rfc3161TimestampTokenInfo TokenInfo { get; private set; }
+
+        /// <summary>
+        /// Get a SignedCms representation of the RFC3161 Timestamp Token.
+        /// </summary>
+        /// <returns>The SignedCms representation of the RFC3161 Timestamp Token.</returns>
+        /// <remarks>
+        /// Successive calls to this method return the same object.
+        /// The SignedCms class is mutable, but changes to that object are not reflected in the
+        /// <see cref="Rfc3161TimestampToken"/> object which produced it.
+        /// The value from calling <see cref="SignedCms.Encode"/> can be interpreted again as an
+        /// <see cref="Rfc3161TimestampToken"/> via another call to <see cref="TryParse"/>.
+        /// </remarks>
+        public SignedCms AsSignedCms() => _parsedDocument;
+
+        public bool VerifyData(ReadOnlySpan<byte> data)
+        {
+            HashAlgorithmName hashAlgorithmName = Helpers.GetDigestAlgorithm(TokenInfo.HashAlgorithmId);
+
+            using (IncrementalHash hasher = IncrementalHash.CreateHash(hashAlgorithmName))
+            {
+                hasher.AppendData(data);
+
+                // SHA-2-512 is the biggest hash we currently know about.
+                Span<byte> stackSpan = stackalloc byte[512 / 8];
+
+                if (hasher.TryGetHashAndReset(stackSpan, out int bytesWritten))
+                {
+                    return VerifyHash(stackSpan.Slice(0, bytesWritten));
+                }
+
+                // Something we understood, but is bigger than 512-bit.
+                // Allocate at runtime, trip in a debug build so we can re-evaluate this.
+                Debug.Fail(
+                    $"TryGetHashAndReset did not fit in {stackSpan.Length} for hash {TokenInfo.HashAlgorithmId.Value}");
+                return VerifyHash(hasher.GetHashAndReset());
+            }
+        }
+
+        public bool VerifyHash(ReadOnlySpan<byte> hash)
+        {
+            return hash.SequenceEqual(TokenInfo.GetMessageHash().Span);
```

Comment:
Nit: Perhaps add a comment that it's ok to use a non-constant time comparer here.